### PR TITLE
Upgrade BDInfo plugin to UHD/Atmos/DTS:X support

### PR DESCRIPTION
**Pull request by stanionascu**
For playback of 1:1 backups via jellyfin, it is required to support not only normal BluRay BDMV folder(s) but also UHD-BD folders. It also keeps the DiscUtils dependency from upstream, so that media info from ISO backups could be extracted.

To make it easier to port later on, the changes from upstream UniqProject/BDInfo are kept to bare minimum. Hence a lot of reverted "fixes".

**Changes**
* Import latest BDInfo changes from https://github.com/UniqProject/BDInfo
* Adds support for UHD-BD BDMV folders streaming
* Keeps ISO DiscUtils support from upstream.

**Issues**
N/A

### CODE REVIEW

**Comment by JustAMan** [comment]
Our current idea is to use as much 3rd party code as possible _without_ including the source.
Can you update the code by _removing_ BDInfo source and instead adding a NuGet dependency on that?

**Comment by stanionascu** [comment]
Makes sense. I'll give it a try.

**Comment by KerryRJ** [comment]
I would like to know what is meant by "adding a NuGet dependency on that"?
Are you referring to using an upstream UniqProject/BDInfo package, that does not seem to exist?

**Comment by dkanada** [comment]
He probably means use the BDInfo library from Nuget instead of committing it here.

**Comment by KerryRJ** [comment]
The only BdInfo Nuget package that I can see in Nuget is the Emby package. Is there another one?

**Comment by stanionascu** [comment]
> The only BdInfo Nuget package that I can see in Nuget is the Emby package. Is there another one?

I already started looking into creating the Nuget package for the updated BDInfo, but it will still require modifications on top, due to usage of `FileSystemMetadata`

**Comment by stanionascu** [comment]
I've stripped out BDInfo into a separate library and published it to nuget repo. It worked just fine with the original UI, but I haven't tested it fully with jellyfin yet.

File: MediaBrowser.MediaEncoding/BdInfo/BdInfoDirectoryInfo.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using BDInfo.IO;
+using MediaBrowser.Model.IO;
+
+namespace MediaBrowser.MediaEncoding.BdInfo
+{
+    class BdInfoDirectoryInfo : BDInfo.IO.IDirectoryInfo
+    {
+        IFileSystem _fileSystem = null;
```
Comment (JustAMan) [inline MediaBrowser.MediaEncoding/BdInfo/BdInfoDirectoryInfo.cs]: what use has this `IFileSystem`? I was under impression we removed most of those abstractions in other places

Comment (stanionascu) [reply MediaBrowser.MediaEncoding/BdInfo/BdInfoDirectoryInfo.cs]: > Mostly looks good.
> 
> What package targets did you publish on Nuget? Got a link?

The package is published to https://www.nuget.org/packages/BDInfo/0.7.6.1 there's only the release BDInfo.dll for netstandard2.1 and netcoreapp3.0.

Comment (stanionascu) [reply MediaBrowser.MediaEncoding/BdInfo/BdInfoDirectoryInfo.cs]: > what use has this `IFileSystem`? I was under impression we removed most of those abstractions in other places

It seems to be still heavily used. This `_fileSystem` is needed for accesing the VFS which Jellyfin in exposing. (e.g. listing files/dirs, opening the filestreams).

**Comment by anthonylavado** [comment]
Okay, for reference @stanionascu has created this NuGet Package here: https://www.nuget.org/packages/BDInfo/0.7.6.1

**Comment by anthonylavado** [comment]
@stanionascu Are you open to transferring the package to us in the Jellyfin organization?

**Comment by stanionascu** [comment]
> @stanionascu Are you open to transferring the package to us in the Jellyfin organization?

Sure. I have no problem with that.

**Comment by JustAMan** [comment]
Do you want to transfer your repo, or want us to fork yours?

**Comment by stanionascu** [comment]
> Do you want to transfer your repo, or want us to fork yours?

I'm fine with transferring.

**Comment by anthonylavado** [comment]
@stanionascu If you go in to the settings, and transfer it to me, I can then transfer it to the organization.

On NuGet, if you go to "Manage Package", you can add "Jellyfin" as an owner. If that doesn't work, add "anthonylavado" as an owner instead.

**Comment by stanionascu** [comment]
> @stanionascu If you go in to the settings, and transfer it to me, I can then transfer it to the organization.
> 
> On NuGet, if you go to "Manage Package", you can add "Jellyfin" as an owner. If that doesn't work, add "anthonylavado" as an owner instead.

I've added Jellyfin as nuget package owner, but I can't transfer the ownership due to missing write permissions to the repos. Feel free to fork it instead.

**Comment by anthonylavado** [comment]
Repo has been forked and added to the org. Package transfer on NuGet.org complete.

**Comment by EraYaN** [comment]
We might want to unlink the fork relation if we want search on the repo.

**Comment by anthonylavado** [comment]
>We might want to unlink the fork relation if we want search on the repo. (Or ask support to do it)

@EraYaN I can ask support to do it. I don’t think you can do it manually.

File: MediaBrowser.MediaEncoding/BdInfo/BdInfoDirectoryInfo.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using BDInfo.IO;
+using MediaBrowser.Model.IO;
+
+namespace MediaBrowser.MediaEncoding.BdInfo
+{
+    class BdInfoDirectoryInfo : BDInfo.IO.IDirectoryInfo
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/BdInfo/BdInfoDirectoryInfo.cs]:
```suggestion
    public class BdInfoDirectoryInfo : BDInfo.IO.IDirectoryInfo
```

**Pull request merged**